In [2]:
import pandas as pd

In [3]:
DATASET_FILE = 'credit_risk_dataset.csv'
credit_risk_data = pd.read_csv(DATASET_FILE)

In [4]:
credit_risk_data.info()

<class 'pandas.DataFrame'>
RangeIndex: 32581 entries, 0 to 32580
Data columns (total 12 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   person_age                  32581 non-null  int64  
 1   person_income               32581 non-null  int64  
 2   person_home_ownership       32581 non-null  str    
 3   person_emp_length           31686 non-null  float64
 4   loan_intent                 32581 non-null  str    
 5   loan_grade                  32581 non-null  str    
 6   loan_amnt                   32581 non-null  int64  
 7   loan_int_rate               29465 non-null  float64
 8   loan_status                 32581 non-null  int64  
 9   loan_percent_income         32581 non-null  float64
 10  cb_person_default_on_file   32581 non-null  str    
 11  cb_person_cred_hist_length  32581 non-null  int64  
dtypes: float64(3), int64(5), str(4)
memory usage: 3.0 MB


In [68]:
credit_risk_data.describe()

,person_age,person_income,person_emp_length,loan_amnt,loan_int_rate,loan_status,loan_percent_income,cb_person_cred_hist_length,range_of_values
count,32581.000000,3.258100e+04,31686.000000,32581.000000,29465.000000,32581.000000,32581.000000,32581.000000,0.0
mean,27.734600,6.607485e+04,4.789686,9589.371106,11.011695,0.218164,0.170203,5.804211,NaN
std,6.348078,6.198312e+04,4.142630,6322.086646,3.240459,0.413006,0.106782,4.055001,NaN
min,20.000000,4.000000e+03,0.000000,500.000000,5.420000,0.000000,0.000000,2.000000,NaN
25%,23.000000,3.850000e+04,2.000000,5000.000000,7.900000,0.000000,0.090000,3.000000,NaN
50%,26.000000,5.500000e+04,4.000000,8000.000000,10.990000,0.000000,0.150000,4.000000,NaN
75%,30.000000,7.920000e+04,7.000000,12200.000000,13.470000,0.000000,0.230000,8.000000,NaN
max,144.000000,6.000000e+06,123.000000,35000.000000,23.220000,1.000000,0.830000,30.000000,NaN


In [69]:
credit_risk_data.mean(numeric_only=True)

person_age                       27.734600
person_income                 66074.848470
person_emp_length                 4.789686
loan_amnt                      9589.371106
loan_int_rate                    11.011695
loan_status                       0.218164
loan_percent_income               0.170203
cb_person_cred_hist_length        5.804211
range_of_values                        NaN
dtype: float64

In [ ]:
# Mean is the sum of all values from the list, devided over the number of elements(length) of that actual list

In [70]:
credit_risk_data.mean(numeric_only=True)

person_age                       27.734600
person_income                 66074.848470
person_emp_length                 4.789686
loan_amnt                      9589.371106
loan_int_rate                    11.011695
loan_status                       0.218164
loan_percent_income               0.170203
cb_person_cred_hist_length        5.804211
range_of_values                        NaN
dtype: float64

In [8]:
# Median is a new statistical measure, in which we find the middle value of a series.
# We have 2 cases that we have to be careful on, if the array length is odd or even.
# In case the length of array is odd, we can find the middle value through this formula: len(arr)/2
# In case the length of the arr is even, we need to calculate the average of the two elements found in the middle

In [71]:
credit_risk_data.median(numeric_only=True)

person_age                       26.00
person_income                 55000.00
person_emp_length                 4.00
loan_amnt                      8000.00
loan_int_rate                    10.99
loan_status                       0.00
loan_percent_income               0.15
cb_person_cred_hist_length        4.00
range_of_values                    NaN
dtype: float64

In [ ]:
# Mode is another statistical measure in which we see the most frequent value in a series.
# In case we have a dataframe, .mode() will output 2 values since we have rows and columns.

In [72]:
credit_risk_data.mode(numeric_only=True)

,person_age,person_income,person_emp_length,loan_amnt,loan_int_rate,loan_status,loan_percent_income,cb_person_cred_hist_length,range_of_values
0,23,60000,0.0,10000,7.51,0,0.1,2,NaN


In [ ]:
# Another important statistical measure is standard deviation - it tells us the distance from each point to the average of the series (aka the spread)

In [73]:
credit_risk_data.std(numeric_only=True)

person_age                        6.348078
person_income                 61983.119168
person_emp_length                 4.142630
loan_amnt                      6322.086646
loan_int_rate                     3.240459
loan_status                       0.413006
loan_percent_income               0.106782
cb_person_cred_hist_length        4.055001
range_of_values                        NaN
dtype: float64

In [ ]:
# Variance is another cool measure, in which we get to know again the spread of the points from the average.
# So then what is the difference? Well, standard deviation is the sqrt of variance.
# Interesting fact is that, variance is widely used to assist in the statistical math operations and its not quite intuitive to understand it as std.
# Standard deviation keeps the same measure unit as the original data.
# meaning, if we have a list of people having height of around 170cm, and the std is 5, then we can see that its ok to see people that have 165cm or 175 cm in height.

In [74]:
credit_risk_data.var(numeric_only=True)

person_age                    4.029810e+01
person_income                 3.841907e+09
person_emp_length             1.716138e+01
loan_amnt                     3.996878e+07
loan_int_rate                 1.050058e+01
loan_status                   1.705737e-01
loan_percent_income           1.140234e-02
cb_person_cred_hist_length    1.644303e+01
range_of_values                        NaN
dtype: float64

In [75]:
credit_risk_data.max(numeric_only=True) - credit_risk_data.min(numeric_only=True)

person_age                        124.00
person_income                 5996000.00
person_emp_length                 123.00
loan_amnt                       34500.00
loan_int_rate                      17.80
loan_status                         1.00
loan_percent_income                 0.83
cb_person_cred_hist_length         28.00
range_of_values                      NaN
dtype: float64

In [76]:
credit_risk_data.corr(numeric_only=True)['loan_status']
credit_risk_data['loan_percent_income'].max()

np.float64(0.83)

## Column-by-column review

The dataset has numerical and categorical columns. The descriptive statistics above cover numerical columns; the tables below summarize every categorical column with counts and proportions. Missing values are also reported so the summaries can be interpreted in context.


In [4]:
categorical_columns = credit_risk_data.select_dtypes(exclude="number").columns

categorical_summary = pd.concat(
    {
        column: pd.DataFrame({
            "count": credit_risk_data[column].value_counts(dropna=False),
            "proportion": credit_risk_data[column].value_counts(dropna=False, normalize=True),
        })
        for column in categorical_columns
    },
    names=["column", "category"],
)
categorical_summary


count  proportion
column                    category                            
person_home_ownership     RENT               16446    0.504773
                          MORTGAGE           13444    0.412633
                          OWN                 2584    0.079310
                          OTHER                107    0.003284
loan_intent               EDUCATION           6453    0.198060
                          MEDICAL             6071    0.186336
                          VENTURE             5719    0.175532
                          PERSONAL            5521    0.169455
                          DEBTCONSOLIDATION   5212    0.159971
                          HOMEIMPROVEMENT     3605    0.110647
loan_grade                A                  10777    0.330776
                          B                  10451    0.320770
                          C                   6458    0.198214
                          D                   3626    0.111292
                          E                    964    0.029588
                          F                    241    0.007397
                          G                     64    0.001964
cb_person_default_on_file N                  26836    0.823670
                          Y                   5745    0.176330

In [5]:
column_quality = pd.DataFrame({
    "dtype": credit_risk_data.dtypes.astype(str),
    "missing_count": credit_risk_data.isna().sum(),
    "missing_percent": credit_risk_data.isna().mean().mul(100).round(2),
    "unique_values": credit_risk_data.nunique(dropna=False),
})
column_quality


,dtype,missing_count,missing_percent,unique_values
person_age,int64,0,0.00,58
person_income,int64,0,0.00,4295
person_home_ownership,str,0,0.00,4
person_emp_length,float64,895,2.75,37
loan_intent,str,0,0.00,6
loan_grade,str,0,0.00,7
loan_amnt,int64,0,0.00,753
loan_int_rate,float64,3116,9.56,349
loan_status,int64,0,0.00,2
loan_percent_income,float64,0,0.00,77


## Absolute correlation with the target

The dataset target is `loan_status`. The next cell ranks numeric columns by the absolute value of their Pearson correlation with that target. The target itself is removed from the ranking.


In [6]:
target_column = "loan_status"
target_correlations = (
    credit_risk_data.corr(numeric_only=True)[target_column]
    .drop(target_column)
    .sort_values(key=lambda correlations: correlations.abs(), ascending=False)
)
target_correlations.to_frame(name="correlation_with_loan_status")


,correlation_with_loan_status
loan_percent_income,0.379366
loan_int_rate,0.335133
person_income,-0.144449
loan_amnt,0.105376
person_emp_length,-0.082489
person_age,-0.021629
cb_person_cred_hist_length,-0.015529


### Interpretation

The strongest numerical associations in this dataset are `loan_percent_income` (about **0.379**) and `loan_int_rate` (about **0.335**). Both are positive: higher values tend to occur alongside `loan_status = 1`. If `1` denotes a default, the first relationship is plausible because a larger loan burden relative to income can make repayment harder. The interest rate may also reflect risk pricing: lenders can charge higher rates to borrowers they assess as riskier, and higher rates also increase the repayment burden.


These explanations are consistent with empirical research on consumer and peer-to-peer lending. For example, a study of LendingClub loans found that borrower indebtedness and interest rates were associated with default risk ([Duarte, Siegel & Young, 2015](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0139427)). That study uses a different lending platform and dataset, so it provides context rather than proof about this dataset.


Pearson correlation measures linear association, not causation. The interest rate may partly encode a lender’s existing risk assessment, and the features may be related to one another. Correlation also covers numeric columns only; it does not measure the relationship of the categorical columns to the target.
